### ISSUE TITLE
the character movements start to slow down

### ISSUE THREAD

**Issue by WK-GiHu** [Linch1/platofrmer#1]
@Linch1

> the character movements start to slow down, when the player is at the center of the background image the lag is more intense and at the ends less intense, I think that the lag is bound to the scroll of the canvas view but don't know what to do for solve it

Every `canvas.bind("<B1-Motion>", scroll_move)` fires a `"<Configure>"` event which forces `resize_window(...)`. 

Try to **allow** `resize_window(...)` only every **10th** event.

Another posibility are, make `resize_window(...)` async from `"<B1-Motion>"` events.
Therefore, `resize_window(...)` have to run in a `.after(xx, resize_window` loop.
With `xx` you set the response time in miliseconds.

**Comment by Linch1** [comment]
Thanks for the response, I tried to remove from the code `canvas.bind("<B1-Motion>", scroll_move)` and
`canvas.bind("<ButtonPress-1>", scroll_start)`  for see if that was the problem but the lag did not decrease. The lag starts when the character move left or right, if it only jumps there is no lag so i think that the problem is with the scroll of the `canvas view` for have always that character at the "camera" center but i don't figure out, i tried to run the code on a better pc with intel core i7 and it seemed to be run with less lag, on my pc i have intel corei3 so can be it relative to the processor?

**Comment by WK-GiHu** [comment]
@Linch1 

>  I tried to remove from the code canvas.bind("<B1-Motion>", scroll_move) and
canvas.bind("<ButtonPress-1>", scroll_start) for see if that was the problem but the lag did not decrease.

Moving with keys, behave the same as i assumed moving with cursor.
On every key press, you do, e.g. `def scroll_screen_left(` and inside `def scroll_screen_left(` you do `canvas.xview_moveto(` whitch fires an `"<Configure>"` event doing `canvas.xview_moveto(` again.
Put a `print('Configure')` into `resize_window(...)` to get a feeling how oft it get called.

> I tried to run the code on a better pc with intel core i7 and it seemed to be run with less lag, on my pc i have intel corei3 so can be it relative to the processor?

Sure, it depends on your Grafiks thrououtput.

**Comment by Linch1** [comment]
i put `print(Counter)` bound to a global variable `Counter` and updating it each time i call the function `resize_window(...)` but it print something only when i move the window or i resize it. When i move the character it prints nothing

**Comment by Linch1** [comment]
Here u can find a shorter executable exaple
https://stackoverflow.com/questions/54991955/tkinter-canvas-image-lag

**Comment by WK-GiHu** [comment]
@Linch1

>  it prints something only when i move the window in the monitor or i resize it.

OK, lets focus at

     if ARROW_LEFT in key_list:
          player.facing_left()
          player.update_image()
          player.move_left()
          scroll_screen_left(player, canvas)


>  i removed from the code the scroll of the camera for have alaways the character centered on the camera and the lag stopped 

This is `scroll_screen_left(player, canvas)`?

**Comment by Linch1** [comment]
yeah if i remove the line  `753 scroll_screen_left(player, canvas)` and the 
line `763 scroll_screen_right(player, canvas)`  the lag stops so theoretically this functions cause the lag but i don't think that is totally a code problem, maybe the background image is only too big (3600x1070)px 

the def of the functions is there
https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L672

**Comment by WK-GiHu** [comment]
@Linch1 

> line 763 scroll_screen_right(player, canvas) the lag stops so theoretically this functions cause the lag 

The only *action* are `canvas.xview_scroll(-to_who.x_speed(), UNITS)`.  
Put a `print(...` into this function, to see how often this get called with which value of `to_who.x_speed()`.

>  the background image is only too big and its scroll causes the lag

This could be verified if you make the main window smaller, try half the Screen size.

---

Verified the `canvas` and `scrollregion`:

https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L63

    total_width=8000, total_height=4000

https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L144

     ws = master.winfo_screenwidth() <= 1800
     hs = master.winfo_screenheight() <= 1070
     Result image size == (3600x1070)px

https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L840

    start_screen_width=700, start_screen_height=700
    Result canvas.size == 700x700
    Changed on '<Configure>' to `.winfo_screenwidth()x.winfo_screenheight()`

https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L841

     scrollregion.bbox == 0, 0, 8000, 4000

Please confirm, if I got the sizes correct.
I think, the image should become equal size of the `scrollregion`, which ist 8000x4000.
Did you move/resize the Background image somewhere?

Also, worth a try, increase `yscrollincrement='1', xscrollincrement='1'` in a range of `2 - 10`

---

https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L581

The async `.after(...` in `player.move_left()`, will slowdown the following `scroll_screen_left(player, canvas)`. Worth a try, move `scroll_screen_left(player, canvas)` at top of `if ARROW_LEFT in key_list:`

**Comment by Linch1** [comment]
i resize the image for have `background width = 2 times the monitor width(3600) ` and 
`background  height like the monitor height(1070)` here
https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L144

the other sizes are the ones you said.
Putting as background a smaller image seems to give no problems, i hope that it will no give problems with more background images toghether, tommorrow i will try, i tried with the things that u said but nothing changed, i assume that the image was only too big..

Thanks for the help :) if u have new ideas for what could generate the problems tell me

**Comment by WK-GiHu** [comment]
@Linch1 

> Putting as background a smaller image seems to give no problems

Think about lower resolution, colors and only zooming instead of resizing.
But, gives lower image quality.

>  i tried with the things that u said but nothing changed

https://github.com/Linch1/platofrmer/blob/80e63d499e7c9108dd130f955a7fbedccecf91c1/game_red/game_.py#L581

I could imagin, that this async loop are still running at the time a **new** call to `player.move_left()` happens.
This results in multiple loops are running.

**Comment by Linch1** [comment]
Yeah i didn’t see that, thanks that can cause other problems.

**Comment by WK-GiHu** [comment]
@Linch1 

Another idea, about image size.
Do it like *Google maps*, tile your image, having only the relevant tiles, visible in the `game_screen_xy`, loaded in the `Canvas`.

**Comment by Linch1** [comment]
what do you mean?

**Comment by WK-GiHu** [comment]
@Linch1 Updated my comment about "background tiles" idea.

**Comment by Linch1** [comment]
oh ok I understood thanks

**Comment by Linch1** [comment]
i'm having another little issue,
at this line i have created the function `change_image(...)`
https://github.com/Linch1/platofrmer/blob/4fbafd3afa3dfad44b58ec4e1d3c94b95a4b8192/game_red/functions.py#L878
that changes gradually an image to another, I've done this becouse i'm trying to create a biomes system,
so when the player passes the "line" that separates two biomes the image start to change gradually. The problem is that when i call this function here

https://github.com/Linch1/platofrmer/blob/4fbafd3afa3dfad44b58ec4e1d3c94b95a4b8192/game_red/functions.py#L128

(this line is bound to this other line)

https://github.com/Linch1/platofrmer/blob/4fbafd3afa3dfad44b58ec4e1d3c94b95a4b8192/game_red/functions.py#L179

 it's like it blocks all the new inputs that i want give so i can't move the player or do other things for a few seconds while the image is changing

**Comment by Linch1** [comment]
i'm having another little issue,
at this line i have created the function `change_image(...)`
https://github.com/Linch1/platofrmer/blob/473873be5002fa145b229794ca861c6b72fcf126/game_red/functions.py#L898
that changes gradually an image to another, I'm trying to follow your advice so i'm trying create a biome system.
I devided the canvas "world" in zones and I assigned at each zone a biome name and image,
so when the player passes the "line" that separates two biomes the image start to change gradually. 
The problem is that when i call this function here

https://github.com/Linch1/platofrmer/blob/473873be5002fa145b229794ca861c6b72fcf126/game_red/functions.py#L121

 it's like it blocks all the new inputs that i want give so i can't move the player or do 
other things for a few seconds while the image is changing.

(I'm working with 3 backgrounds toghether so i have to change 3 images at the same time. I don't thing this is the problem becouse if i work with only 1 background it gives the same issue but for les time)

**Comment by Linch1** [comment]
here you can find and executable example for the issue https://stackoverflow.com/questions/55080071/tkinter-chage-gradually-two-images-make-lag
